File: backend/app/routers/billing.py
"""Stripe billing — checkout, webhook, portal."""

import stripe
from fastapi import APIRouter, Depends, HTTPException, Request, status
from fastapi.responses import RedirectResponse, JSONResponse
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select

from app.config import settings
from app.database import get_db
from app.deps import get_current_user
from app.models import User, Plan

router = APIRouter(prefix="/billing", tags=["billing"])


def _stripe_client() -> stripe.StripeClient:
    if not settings.STRIPE_SECRET_KEY:
        raise HTTPException(status_code=503, detail="Billing not configured")
    return stripe.StripeClient(settings.STRIPE_SECRET_KEY)


PLAN_PRICE_MAP = {
    "pro": lambda: settings.STRIPE_PRICE_PRO,
    "business": lambda: settings.STRIPE_PRICE_BUSINESS,
}


@router.post("/checkout/{plan}")
async def create_checkout(
    plan: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Create a Stripe Checkout Session and return the redirect URL."""
    plan = plan.lower()
    if plan not in PLAN_PRICE_MAP:
        raise HTTPException(status_code=400, detail="Invalid plan. Use 'pro' or 'business'.")

    price_id = PLAN_PRICE_MAP[plan]()
    if not price_id:
        raise HTTPException(status_code=503, detail="Plan pricing not configured")

    client = _stripe_client()

    # Create / reuse Stripe customer
    customer_id = current_user.stripe_customer_id
    if not customer_id:
        customer = client.customers.create(
            params={"email": current_user.email, "metadata": {"user_id": str(current_user.id)}}
        )
        customer_id = customer.id
        current_user.stripe_customer_id = customer_id
        await db.commit()

    session = client.checkout.sessions.create(
        params={
            "customer": customer_id,
            "mode": "subscription",
            "line_items": [{"price": price_id, "quantity": 1}],
            "success_url": f"{settings.APP_URL}/dashboard?upgraded=1",
            "cancel_url": f"{settings.APP_URL}/pricing",
            "metadata": {"user_id": str(current_user.id), "plan": plan},
        }
    )
    return {"url": session.url}


@router.get("/portal")
async def billing_portal(
    current_user: User = Depends(get_current_user),
):
    """Redirect to Stripe Customer Portal for subscription management."""
    if not current_user.stripe_customer_id:
        raise HTTPException(status_code=400, detail="No active subscription found")

    client = _stripe_client()
    session = client.billing_portal.sessions.create(
        params={
            "customer": current_user.stripe_customer_id,
            "return_url": f"{settings.APP_URL}/dashboard",
        }
    )
    return {"url": session.url}


@router.post("/webhook")
async def stripe_webhook(request: Request, db: AsyncSession = Depends(get_db)):
    """Handle Stripe webhook events."""
    payload = await request.body()
    sig_header = request.headers.get("stripe-signature", "")

    if not settings.STRIPE_WEBHOOK_SECRET:
        raise HTTPException(status_code=503, detail="Webhook not configured")

    try:
        event = stripe.Webhook.construct_event(
            payload, sig_header, settings.STRIPE_WEBHOOK_SECRET
        )
    except stripe.error.SignatureVerificationError:
        raise HTTPException(status_code=400, detail="Invalid signature")

    event_type = event["type"]

    if event_type == "checkout.session.completed":
        session = event["data"]["object"]
        user_id = session.get("metadata", {}).get("user_id")
        plan_name = session.get("metadata", {}).get("plan", "pro")
        customer_id = session.get("customer")

        if user_id:
            result = await db.execute(select(User).where(User.id == user_id))
            user = result.scalar_one_or_none()
            if user:
                user.plan = Plan(plan_name)
                if customer_id:
                    user.stripe_customer_id = customer_id
                await db.commit()

    elif event_type in ("customer.subscription.updated", "customer.subscription.deleted"):
        subscription = event["data"]["object"]
        customer_id = subscription.get("customer")
        sub_status = subscription.get("status")

        result = await db.execute(
            select(User).where(User.stripe_customer_id == customer_id)
        )
        user = result.scalar_one_or_none()

        if user:
            if event_type == "customer.subscription.deleted" or sub_status in ("canceled", "unpaid", "incomplete_expired"):
                user.plan = Plan.FREE
            elif sub_status == "active":
                # Re-derive plan from the price ID
                items = subscription.get("items", {}).get("data", [])
                if items:
                    price_id = items[0].get("price", {}).get("id", "")
                    if price_id == settings.STRIPE_PRICE_BUSINESS:
                        user.plan = Plan.BUSINESS
                    elif price_id == settings.STRIPE_PRICE_PRO:
                        user.plan = Plan.PRO
            await db.commit()

    return JSONResponse({"received": True})

File: worker/src/routes/billing.ts
import { Hono } from "hono";
import Stripe from "stripe";
import { eq } from "drizzle-orm";
import { getDb } from "../db/client";
import { users } from "../db/schema";
import { authRequired, type AuthUser } from "../middleware/auth";
import type { Env } from "../lib/types";

const app = new Hono<{ Bindings: Env }>();

function getStripe(env: Env) {
  if (!env.STRIPE_SECRET_KEY) throw new Error("Billing not configured");
  return new Stripe(env.STRIPE_SECRET_KEY, { apiVersion: "2024-12-18.acacia" as any });
}

/** POST /api/billing/checkout/:plan */
app.post("/checkout/:plan", authRequired, async (c) => {
  const plan = c.req.param("plan").toLowerCase();
  const user = c.get("user") as AuthUser;

  const priceMap: Record<string, string> = {
    pro: c.env.STRIPE_PRICE_PRO,
    business: c.env.STRIPE_PRICE_BUSINESS,
  };

  const priceId = priceMap[plan];
  if (!priceId) {
    return c.json({ detail: "Invalid plan. Use 'pro' or 'business'." }, 400);
  }

  const stripe = getStripe(c.env);
  const db = getDb(c.env.DATABASE_URL);

  let customerId = user.stripeCustomerId;
  if (!customerId) {
    const customer = await stripe.customers.create({
      email: user.email,
      metadata: { user_id: user.id },
    });
    customerId = customer.id;
    await db.update(users).set({ stripeCustomerId: customerId }).where(eq(users.id, user.id));
  }

  const session = await stripe.checkout.sessions.create({
    customer: customerId,
    mode: "subscription",
    line_items: [{ price: priceId, quantity: 1 }],
    success_url: `${c.env.APP_URL}/dashboard?upgraded=1`,
    cancel_url: `${c.env.APP_URL}/pricing`,
    metadata: { user_id: user.id, plan },
  });

  return c.json({ url: session.url });
});

/** GET /api/billing/portal */
app.get("/portal", authRequired, async (c) => {
  const user = c.get("user") as AuthUser;
  if (!user.stripeCustomerId) {
    return c.json({ detail: "No active subscription found" }, 400);
  }

  const stripe = getStripe(c.env);
  const session = await stripe.billingPortal.sessions.create({
    customer: user.stripeCustomerId,
    return_url: `${c.env.APP_URL}/dashboard`,
  });

  return c.json({ url: session.url });
});

/** POST /api/billing/webhook */
app.post("/webhook", async (c) => {
  const stripe = getStripe(c.env);
  const sig = c.req.header("stripe-signature");
  if (!sig) return c.json({ detail: "Missing signature" }, 400);

  const body = await c.req.text();
  let event: Stripe.Event;
  try {
    event = stripe.webhooks.constructEvent(body, sig, c.env.STRIPE_WEBHOOK_SECRET);
  } catch {
    return c.json({ detail: "Invalid signature" }, 400);
  }

  const db = getDb(c.env.DATABASE_URL);

  if (event.type === "checkout.session.completed") {
    const session = event.data.object as Stripe.Checkout.Session;
    const userId = session.metadata?.user_id;
    const plan = session.metadata?.plan?.toUpperCase() as "PRO" | "BUSINESS";
    if (userId && plan) {
      await db.update(users).set({ plan }).where(eq(users.id, userId));
    }
  }

  if (event.type === "customer.subscription.updated") {
    const sub = event.data.object as Stripe.Subscription;
    if (sub.status === "active") {
      const customer = await stripe.customers.retrieve(sub.customer as string);
      if ("metadata" in customer && customer.metadata?.user_id) {
        // Plan already set at checkout, but update if needed
      }
    }
  }

  if (event.type === "customer.subscription.deleted") {
    const sub = event.data.object as Stripe.Subscription;
    const customer = await stripe.customers.retrieve(sub.customer as string);
    if ("metadata" in customer && customer.metadata?.user_id) {
      await db.update(users).set({ plan: "FREE" }).where(eq(users.id, customer.metadata.user_id));
    }
  }

  return c.json({ received: true });
});

export default app;
